src/lambda/config/parameterUtils.ts

```diff
@@ -0,0 +1,81 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as path from 'path'
+import * as vscode from 'vscode'
+import { CloudFormation } from '../../shared/cloudformation/cloudformation'
+import { load as loadTemplatesConfig } from './templates'
+
+export interface GetParametersContext {
+    loadTemplate: typeof CloudFormation.load
+}
+
+export async function getParameters(
+    templateUri: vscode.Uri,
+    context: GetParametersContext = { loadTemplate: CloudFormation.load }
+): Promise<Map<string, { required: boolean }>> {
+    const template = await context.loadTemplate(templateUri.fsPath)
+    if (!template.Parameters) {
+        return new Map()
+    }
+
+    const result = new Map<string, { required: boolean }>()
+
+    for (const name of Object.getOwnPropertyNames(template.Parameters)) {
+        const parameter = template.Parameters[name]!
+
+        result.set(name, {
+            // Explicitly compare with undefined, as a valid default value may be falsy.
+            required: parameter.Default === undefined
+        })
+    }
+
+    return result
+}
+
+export async function getParameterNames(
+    templateUri: vscode.Uri,
+    context: GetParametersContext = { loadTemplate: CloudFormation.load }
+) {
+    return [...(await getParameters(templateUri, context)).keys()]
+}
+
+export interface GetOverriddenParametersContext {
+    readonly loadTemplatesConfig: typeof loadTemplatesConfig
+    getWorkspaceFolder(uri: vscode.Uri): Pick<vscode.WorkspaceFolder, 'uri'> | undefined
+}
+
+export class DefaultGetOverriddenParametersContext implements GetOverriddenParametersContext {
+    public readonly getWorkspaceFolder = vscode.workspace.getWorkspaceFolder
+    public readonly loadTemplatesConfig = loadTemplatesConfig
+}
+
+export async function getOverriddenParameters(
+    templateUri: vscode.Uri,
+    context: GetOverriddenParametersContext = new DefaultGetOverriddenParametersContext()
+): Promise<Map<string, string> | undefined> {
+    const workspaceFolder = context.getWorkspaceFolder(templateUri)
+    if (!workspaceFolder) {
+        // This should never happen.
+        throw new Error(`The template ${templateUri.fsPath} is not in the workspace`)
+    }
+
+    const relativeTemplatePath = vscode.Uri.file(path.relative(workspaceFolder.uri.fsPath, templateUri.fsPath)).fsPath
```

Comment:
I've got this coming in a change, currently in templates.ts. It is a good idea to consolidate the relative pathing logic, and normalize paths. Please steal this!

```javascript
export function getTemplateRelativePath(
    templatePath: string,
    workspaceFolderPath: string
): string {
    const relativePath = _path.relative(workspaceFolderPath, templatePath)

    return relativePath.replace(
        _path.sep,
        '/'
    )
}
```

Reply:
Added a new file `shared/utilities/pathUtilities` with these functions:

```
export function getNormalizedRelativePath(from: string, to: string): string {
    return normalizeSeparator(_path.relative(from, to))
}

export function normalizeSeparator(path: string) {
    return path.replace(_path.sep, _path.posix.sep)
}
```

Reply:
✔️